File: src/platform/Aspen.Data.ClientQueries/ClientQueryOptions.cs
using System.Text.Json.Serialization;
using Aspen.Data.ClientQueries.Options;

namespace Aspen.Data.ClientQueries;

public class ClientQueryOptions
{
    [JsonPropertyName("filters")]
    public ClientQueryFilterOption[] Filters { get; set; } = [];

    [JsonPropertyName("sorts")]
    public ClientQuerySortOption[] Sorts { get; set; } = [];

    [JsonPropertyName("pagination")]
    public ClientQueryPaginationOption? Pagination { get; set; }
}

public class TransformativeClientQueryOptions : ClientQueryOptions
{

}

File: src/web/src/lib/querying/ClientQueryOptions.ts
import stableHash from "stable-hash";

export interface ClientQueryOptions {
	readonly filters?: ClientQueryFilterOption[];
	readonly sorts?: ClientQuerySortOption[];
	readonly pagination?: ClientQueryPaginationOption;
}

export interface ClientQueryFilterOption {
	readonly column?: string;
	readonly operator: FilterOperator;
	readonly operand?: string | number | boolean;
	readonly filters?: ClientQueryFilterOption[];
}

export interface ClientQuerySortOption {
	readonly column: string;
	readonly direction: SortDirection;
}

export interface ClientQueryPaginationOption {
	readonly cursor: number;
	readonly pageSize: number;
}

export enum SortDirection {
	Ascending = "ASC",
	Descending = "DESC"
}

export enum FilterOperator {
	EqualTo = "==",
	NotEqualTo = "!=",
	GreaterThan = ">",
	GreaterThanOrEqualTo = ">=",
	LessThan = "<",
	LessThanOrEqualTo = "<=",
	And = "&&",
	Or = "||"
}

export const hashClientOptions = (options: ClientQueryOptions): string => {
	return stableHash(options);
};
